pkg/storage/ingest/pusher.go

```diff
@@ -119,16 +136,23 @@ func (c pusherConsumer) unmarshalRequests(ctx context.Context, records []record,
 	defer close(recC)
 	done := ctx.Done()
 
-	for _, record := range records {
+	for _, rec := range records {
+		// rec.ctx contains the tracing baggage for this record, which we propagate down the call tree.
+		// But this context's cancellation is disjointed from the top-most ctx. The context.AfterFunc below,
+		// is what fuses the two lifetimes together.
+		recCtx, cancel := context.WithCancelCause(rec.ctx)
```

Comment:
[nit] I would renaming the cancel function for clarity.

```suggestion
		recCtx, cancelRecCtx := context.WithCancelCause(rec.ctx)
```